sway-ir/src/verify.rs

```diff
@@ -1017,7 +1014,15 @@ impl<'a, 'eng> InstructionVerifier<'a, 'eng> {
     // This is a really common operation above... calling `Value::get_type()` and then failing when
     // two don't match.
     fn opt_ty_not_eq(&self, l_ty: &Option<Type>, r_ty: &Option<Type>) -> bool {
-        l_ty.is_none() || r_ty.is_none() || !l_ty.unwrap().eq(self.context, r_ty.as_ref().unwrap())
+        l_ty.is_none()
+            || r_ty.is_none()
+            || (!l_ty.unwrap().is_never(self.context)
+                && !r_ty.unwrap().is_never(self.context)
+                && !l_ty.unwrap().eq(self.context, r_ty.as_ref().unwrap()))
+    }
+
+    fn ty_not_eq(&self, l_ty: &Type, r_ty: &Type) -> bool {
+        !l_ty.is_never(self.context) && !r_ty.is_never(self.context) && !l_ty.eq(self.context, r_ty)
```

Comment:
The `Type::eq()` is not implemented for `Never` (in _irtype.rs_), that needs to be done. If it were, it looks to me that `ty_not_eq` wouldn't be needed at all. `Never` comparison would also be done in `eq()`.